### ISSUE TITLE
InsertionOrderSet performance issue

### ISSUE THREAD

**Issue by kpbochenek** [scalatest/scalatest#2307]
Hi,

I have a suite where I add dynamically tests, it is ~10k that I noticed heavy slowdown in running our tests.
Investigating what takes so long leads to:
```
  override def testNames: Set[String] = {
    InsertionOrderSet(atomic.get.testNamesList)
  }
```

When I have my Spec and for testing will do inside:
```
override def testNames: Set[String] = {
  log("start time")
  val result = super.testNames
  log("end time")
  result
}
```
It takes minutes to finish.
 
InsertionOrderSet is defined as:
```
  def apply[A](elements: List[A]): Set[A] =
    scala.collection.immutable.SortedSet(elements: _*)(new Ordering[A] {
      def compare(x: A, y: A): Int = elements.indexOf(x) compare elements.indexOf(y)
    })
```
Which if I understand correctly uses O(n) to check indexOf -> causes sorting of `elements: List[]` to be O(n^2*log(n)) ?

I know it is uncommon to have so many tests in a suite but could we somehow add optimization to also handle with reasonable performance suites with bigger amount of tests?

Also I tried to see if I could work around for now in my current version, but I can see it uses `atomic` which is private to spec, do you have any idea if it is possible to do something in my code using current scalatest to speed things up?

~Cheers
Krzysiek

**Comment by cheeseng** [comment]
@kpbochenek Sorry to hear about your case, unfortunately I can't think of a way you can adjust your code to speed things up.  I am working on improving the performance by using a implementation that wraps a scala.collection.mutable.LinkedHashSet underlying, I'll try to include that and release 3.2.18 this week, if the changes looks good to @bvenners .  I should be able to submit the PR today.

**Comment by kpbochenek** [comment]
Thanks for a quick response 👍 

For now I just copy&pasted FlatMapSpecLike and renamed to MyFlatMapSpec which I then used and changed InsertionOrderSet to simple Set and that fixed my performance issue. I don't care about order so its fine in my case but in general it cannot be used :/

Just a reference immutable SeqSet would have a perfect use here: https://github.com/scala/scala-library-next/issues/22

**Comment by cheeseng** [comment]
@kpbochenek Fyi I submitted the PR here: 

https://github.com/scalatest/scalatest/pull/2308

**Comment by kpbochenek** [comment]
@cheeseng awesome thanks for fixing it so fast! 👍

**Comment by cheeseng** [comment]
@kpbochenek fyi the PR is included in 3.2.18 that is already released.
